File: src/download/downloader.rs
use std::sync::Arc;

use anyhow::Result;
use futures::future::join_all;
use prost::Message;
use reqwest::header::{self, CONTENT_RANGE, CONTENT_TYPE};
use reqwest::header::{HeaderMap, ACCEPT_RANGES, CONTENT_LENGTH};
use thiserror::Error;
use tokio::fs;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::sync::Semaphore;

use crate::{replace_illegal_chars_in_windows, util};

use super::{DanmakuSegment, Video};

const API_INFO: &'static str = "https://api.bilibili.com/x/web-interface/view?bvid=";
const API_PLAY: &'static str = "https://api.bilibili.com/x/player/playurl";
const API_BULLET: &'static str = "http://api.bilibili.com/x/v2/dm/web/seg.so";
const API_USERINFO: &'static str = "https://api.bilibili.com/x/web-interface/nav";
const UA: &'static str = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.132 Safari/537.36";

#[derive(Error, Debug)]
pub enum DownloadError<'a> {
    #[error("登陆失败")]
    LoginFail,
    #[error("获取视频信息 {0} 失败")]
    GetVideoInfoFail(&'a str),
    #[error("获取弹幕包失败 cid: {0}, segment: {1}")]
    GetDanmakuSegmentFail(u64, u64),
}

#[derive(Debug, Clone)]
pub struct Downloader {
    client: reqwest::Client,
    task_num: u8,
    pub dir: String,
}

fn extract_content_range(headers: &HeaderMap) -> Result<u64> {
    if let Some(range) = headers.get(CONTENT_RANGE) {
        let _range = String::from(range.to_str()?);
        let mut splits = _range.split("/");
        let _ = splits.next();
        if let Some(len) = splits.next() {
            return Ok(len.parse::<u64>()?);
        }
    }
    Ok(0 as u64)
}

fn extract_content_len(headers: &HeaderMap) -> Result<u64> {
    let check_accept = match headers.get(ACCEPT_RANGES) {
        Some(accept_ranges) => accept_ranges.to_str()?.contains("bytes"),
        None => false,
    };

    let content_length = if check_accept {
        match headers.get(CONTENT_LENGTH) {
            Some(lenth) => lenth.to_str()?.parse::<usize>()?,
            None => 0,
        }
    } else {
        0
    };
    Ok(content_length as u64)
}

fn extract_format(content_type: &str) -> String {
    match content_type {
        "video/mp4" => "mp4",
        "video/x-flv" => "flv",
        "application/x-mpegURL" => "m3u8",
        "video/MP2T" => "ts",
        "video/3gpp" => "3gpp",
        "video/quicktime" => "mov",
        "video/x-msvideo" => "avi",
        "video/x-ms-wmv" => "wmv",
        "audio/x-wav" => "wav",
        "audio/x-mp3" => "mp3",
        "audio/mp4" => "mp4",
        "application/ogg" => "ogg",
        "image/jpeg" => "jpeg",
        "image/png" => "png",
        "image/tiff" => "tiff",
        "image/gif" => "gif",
        "image/svg+xml" => "svg",
        _ => "mp4",
    }
    .to_string()
}

fn extract_format_from_header(headers: &HeaderMap) -> String {
    match headers.get(CONTENT_TYPE) {
        Some(content_type) => extract_format(content_type.to_str().unwrap_or("")),
        None => "mp4".to_string(),
    }
}

async fn request_json(builder: reqwest::RequestBuilder) -> Result<serde_json::Value> {
    Ok(builder.send().await?.json::<serde_json::Value>().await?)
}

impl Downloader {
    pub fn new(task_num: u8) -> Result<Self> {
        let mut headers = header::HeaderMap::new();
        headers.insert(
            "referer",
            header::HeaderValue::from_static("https://www.bilibili.com/"),
        );
        if let Ok(c) = std::fs::read_to_string("cookie.txt") {
            let val = header::HeaderValue::from_str(c.as_str())?;
            headers.insert("cookie", val);
        }
        let client = reqwest::Client::builder()
            .user_agent(UA)
            .default_headers(headers)
            .build()?;
        let dir = "".to_string();
        let downloader = Downloader {
            client,
            task_num,
            dir,
        };
        Ok(downloader)
    }

    /// Validate cookie
    pub async fn check_login(&self) -> Result<()> {
        let r = &request_json(self.client.get(API_USERINFO)).await?;
        match r["data"]["isLogin"].as_bool() {
            Some(true) => {
                println!("login success");
                Ok(())
            }
            _ => Err(DownloadError::LoginFail.into()),
        }
    }

    /// Buile video
    pub async fn build_video(&self, bv: String) -> Result<Video> {
        self.check_login().await?;
        let url = format!("{}{}", API_INFO, bv);
        // println!("video info: {}", url);
        let response = &request_json(self.client.get(url)).await?["data"];
        let title = response["title"].as_str().unwrap_or_else(|| bv.as_str());
        let title = replace_illegal_chars_in_windows(title);
        let cid = response["cid"]
            .as_u64()
            .ok_or_else(|| DownloadError::GetVideoInfoFail("cid"))?;
        let duration = response["duration"]
            .as_u64()
            .ok_or_else(|| DownloadError::GetVideoInfoFail("duration"))?;
        // println!("video play info: {}?bvid={}&cid={}", API_PLAY, bv, cid);
        let response = &request_json(self.client.get(API_PLAY).query(&[
            ("bvid", bv.as_str()),
            ("cid", cid.to_string().as_str()),
            // add this to fetch 1080P or better
            ("fnval", "2000"),
        ]))
        .await?["data"];
        let video_url;
        let audio_url;
        let format;
        let content_len;

        if let Some(data) = response["dash"].as_object() {
            let video_data = &data
                .get("video")
                .ok_or_else(|| DownloadError::GetVideoInfoFail("url"))?[0];
            video_url = video_data["baseUrl"].as_str().unwrap_or("").to_string();
            audio_url = data
                .get("audio")
                .ok_or_else(|| DownloadError::GetVideoInfoFail("url"))?[0]["baseUrl"]
                .as_str()
                .unwrap_or("")
                .to_string();
            format = extract_format(video_data["mimeType"].as_str().unwrap_or(""));
            let response = self
                .client
                .get(&video_url)
                .header("range", "bytes=0-1024")
                .send()
                .await?;
            content_len = extract_content_range(response.headers()).unwrap_or(0);
        } else {
            video_url = response["durl"]
                .as_array()
                .ok_or_else(|| DownloadError::GetVideoInfoFail("durl"))?[0]["url"]
                .as_str()
                .ok_or_else(|| DownloadError::GetVideoInfoFail("url"))?
                .to_string();
            audio_url = String::new();
            let response = self.client.head(&video_url).send().await?;
            format = extract_format_from_header(response.headers());
            content_len = extract_content_len(response.headers()).unwrap_or(0);
        }
        println!(
            "video format: {}, size: {} MB",
            format,
            content_len / (1024 * 1024) as u64
        );
        Ok(Video {
            bv,
            cid,
            video_url,
            audio_url,
            title,
            format,
            duration,
            content_len,
        })
    }

    /// Download video chunks
    pub async fn download_chunks(self: Arc<Self>, video: Arc<Video>) -> Result<u16> {
        // 10MB for one chunk
        let chunk_size = 1024 * 1024 * 10;
        let mut handler_list = vec![];
        let semaphore = Arc::new(Semaphore::new(self.task_num.into()));

        let mut start = 0;
        let mut end = 0;
        let mut index = 0;
        while end < video.content_len {
            end += chunk_size;
            if end > video.content_len {
                end = video.content_len;
            }
            let permit = semaphore.clone().acquire_owned().await.unwrap();
            let downloader = self.clone();
            let video = video.clone();
            let handler = tokio::spawn(async move {
                let _ = downloader
                    .download_chunk(video, (start, end), index as u8)
                    .await;
                drop(permit);
            });
            handler_list.push(handler);
            start = end + 1;
            index += 1;
        }
        join_all(handler_list).await;
        Ok(index)
    }

    /// Download single video chunk
    pub async fn download_chunk(
        self: Arc<Self>,
        video: Arc<Video>,
        range: (u64, u64),
        index: u8,
    ) -> Result<()> {
        println!("download chunk {} from {} to {}", index, range.0, range.1);
        let mut response = self
            .client
            .get(video.video_url.as_str())
            .header("range", format!("bytes={}-{}", range.0, range.1))
            .send()
            .await?;
        let mut offset = 0;
        while let Some(bytes) = response.chunk().await? {
            let filepath = format!("{}/chunk_{}", self.dir, index);
            let len = bytes.len() as u64;
            util::write_bytes_to_file(filepath.as_str(), &bytes, offset).await?;
            offset += len;
        }
        Ok(())
    }

    /// Download audio if exists
    pub async fn download_audio(self: Arc<Self>, video: Arc<Video>) -> Result<()> {
        if video.audio_url.len() == 0 {
            return Ok(());
        }
        let response = self.client.get(video.audio_url.as_str()).send().await?;
        if let Ok(bytes) = response.bytes().await {
            let filepath = format!("{}/audio.mp3", self.dir);
            util::write_bytes_to_file(filepath.as_str(), &bytes, 0).await?;
            println!("downloading audio success at {}", filepath);
        }
        Ok(())
    }

    /// Merge chunk files and mix video and audio
    pub async fn build_final_video(
        self: Arc<Self>,
        video: Arc<Video>,
        chunk_size: u16,
    ) -> Result<()> {
        let format = video.format.as_str();
        let video_path = format!("{}/video.{}", self.dir, format);
        let mut file = util::create_file(video_path.as_str()).await?;

        for index in 0..chunk_size {
            println!("merge chunk file {}", index);
            let chunk_path = format!("{}/chunk_{}", self.dir, index);
            if let Ok(mut chunk_file) = fs::File::open(chunk_path.as_str()).await {
                let size = chunk_file.metadata().await?.len();
                let mut buf = vec![0; size as usize];
                chunk_file.read_exact(&mut buf).await?;
                file.write_all(&buf).await?;
                fs::remove_file(chunk_path).await?;
            }
        }
        println!("merge chunk files finished");

        let video_path = format!("{}/video.{}", self.dir, format);
        let audio_path = format!("{}/audio.mp3", self.dir);
        let output_path = format!("{}/{}.{}", self.dir, video.title.as_str(), format);
        let _ = util::mix_video_audio(
            video_path.as_str(),
            audio_path.as_str(),
            output_path.as_str(),
        )
        .await?;
        fs::remove_file(video_path).await?;
        fs::remove_file(audio_path).await?;
        Ok(())
    }

    /// Download danmaku
    pub async fn download_danmaku(self: Arc<Self>, video: Arc<Video>) -> Result<()> {
        println!("downloading danmaku...");
        let mut file = util::create_file(format!("{}/danmuku.txt", self.dir).as_str()).await?;
        let mut fut = vec![];
        let bags = (video.duration + 359) / 360;
        for i in 0..bags {
            let downloader = self.clone();
            let video = video.clone();
            fut.push(downloader.download_danmaku_segment(video, i + 1));
        }
        let danmuku_list = futures::future::try_join_all(fut).await?;
        for danmuku_seg in danmuku_list {
            for danmuku in danmuku_seg.elems {
                if let Ok(json) = serde_json::to_string(&danmuku) {
                    file.write_all(format!("{}\n", json).as_bytes()).await?;
                }
            }
        }
        Ok(())
    }

    /// Download danmaku segment
    pub async fn download_danmaku_segment(
        self: Arc<Self>,
        video: Arc<Video>,
        seg_index: u64,
    ) -> Result<DanmakuSegment> {
        let response = self
            .client
            .get(API_BULLET)
            .query(&[
                ("oid", video.cid),
                ("segment_index", seg_index),
                ("type", 1),
            ])
            .send()
            .await?;
        let content = response.bytes().await?;
        match super::DanmakuSegment::decode(content) {
            Ok(v) => Ok(v),
            Err(_) => Err(DownloadError::GetDanmakuSegmentFail(video.cid, seg_index).into()),
        }
    }

    /// Start download mission
    pub async fn start(self: Arc<Self>, video: Arc<Video>) -> Result<()> {
        let chunk_size = self.clone().download_chunks(video.clone()).await?;
        self.clone().download_audio(video.clone()).await?;
        self.clone()
            .build_final_video(video.clone(), chunk_size)
            .await?;
        self.clone().download_danmaku(video).await?;
        Ok(())
    }
}

File: bilidown/downloader.py
"""Bilibili 视频下载器核心模块"""

import os

import requests
from fake_useragent import UserAgent

from .model import Video

_ua_pool = UserAgent()


class DownloadError(Exception):
    """下载器异常基类"""
    pass


class LoginFailError(DownloadError):
    """Cookie 登录验证失败"""
    pass


class VideoInfoError(DownloadError):
    """获取视频信息失败"""
    pass


class DanmakuError(DownloadError):
    """弹幕下载/解析失败"""
    pass


class Downloader:
    """Bilibili 视频下载器"""

    API_USERINFO = "https://api.bilibili.com/x/web-interface/nav"
    API_INFO = "https://api.bilibili.com/x/web-interface/view?bvid="
    API_PLAY = "https://api.bilibili.com/x/player/playurl"
    API_BULLET = "http://api.bilibili.com/x/v2/dm/web/seg.so"

    def __init__(self, task_num: int = 7):
        self.task_num = min(task_num, 10)
        self.dir = ""
        self.session = requests.Session()
        self.session.headers.update({
            "User-Agent": _ua_pool.random,
            "Referer": "https://www.bilibili.com/",
        })
        if os.path.exists("cookie.txt"):
            try:
                with open("cookie.txt") as f:
                    self.session.headers["Cookie"] = f.read().strip()
            except OSError as e:
                print(f"读取 cookie.txt 失败: {e}")

    def check_login(self) -> None:
        """验证 Cookie 是否有效"""
        try:
            r = self.session.get(self.API_USERINFO, timeout=10)
            r.raise_for_status()
        except requests.RequestException as e:
            raise LoginFailError(f"登录验证请求失败: {e}")
        if not r.json().get("data", {}).get("isLogin"):
            raise LoginFailError("Cookie 登录验证失败")

    def _extract_format(self, content_type: str) -> str:
        """根据 MIME 类型提取文件格式"""
        mapping = {
            "video/mp4": "mp4", "video/x-flv": "flv",
            "application/x-mpegURL": "m3u8", "video/MP2T": "ts",
            "video/3gpp": "3gpp", "video/quicktime": "mov",
            "video/x-msvideo": "avi", "video/x-ms-wmv": "wmv",
            "audio/x-wav": "wav", "audio/x-mp3": "mp3",
            "audio/mp4": "mp4", "application/ogg": "ogg",
            "image/jpeg": "jpeg", "image/png": "png",
            "image/tiff": "tiff", "image/gif": "gif",
            "image/svg+xml": "svg",
        }
        return mapping.get(content_type, "mp4")

    def build_video(self, bv: str) -> "Video":
        """根据 BV 号构建 Video 对象"""
        from .model import Video
        from .util import replace_illegal_chars_in_windows

        self.check_login()

        # Step 1: 获取视频基本信息
        info_url = f"{self.API_INFO}{bv}"
        try:
            resp = self.session.get(info_url, timeout=10)
            resp.raise_for_status()
            info = resp.json()["data"]
        except requests.RequestException as e:
            raise VideoInfoError(f"请求视频信息失败: {e}")
        except (KeyError, ValueError, TypeError) as e:
            raise VideoInfoError(f"解析视频信息响应失败: {e}")

        title = replace_illegal_chars_in_windows(info.get("title", bv))
        cid = info["cid"]
        duration = info["duration"]

        # Step 2: 获取播放地址
        try:
            resp = self.session.get(
                self.API_PLAY,
                params={"bvid": bv, "cid": cid, "fnval": "2000"},
                timeout=10,
            )
            resp.raise_for_status()
            play = resp.json()["data"]
        except requests.RequestException as e:
            raise VideoInfoError(f"请求播放地址失败: {e}")
        except (KeyError, ValueError, TypeError) as e:
            raise VideoInfoError(f"解析播放地址响应失败: {e}")

        # Step 3: 解析 DASH 或 FLV
        if "dash" in play and play["dash"] is not None:
            video_data = play["dash"]["video"][0]
            video_url = video_data["baseUrl"]
            audio_url = play["dash"]["audio"][0]["baseUrl"]
            fmt = self._extract_format(video_data.get("mimeType", ""))

            r = self.session.get(video_url, headers={"Range": "bytes=0-1024"})
            content_range = r.headers.get("Content-Range", "")
            content_len = int(content_range.split("/")[-1]) if "/" in content_range else 0
        else:
            durl = play["durl"][0]
            video_url = durl["url"]
            audio_url = ""
            r = self.session.head(video_url)
            fmt = self._extract_format(r.headers.get("Content-Type", ""))
            content_len = int(r.headers.get("Content-Length", 0))

        return Video(
            bv=bv,
            cid=cid,
            video_url=video_url,
            audio_url=audio_url,
            title=title,
            format=fmt,
            duration=duration,
            content_len=content_len,
        )

    def download_chunk(self, video: "Video", range_tuple: tuple, index: int) -> None:
        """下载单个视频分块"""
        from .util import write_bytes_to_file

        start, end = range_tuple
        r = self.session.get(
            video.video_url,
            headers={"Range": f"bytes={start}-{end}"},
            stream=True,
            timeout=30,
        )
        r.raise_for_status()
        filepath = f"{self.dir}/chunk_{index}"
        offset = 0
        for chunk in r.iter_content(chunk_size=8192):
            if chunk:
                write_bytes_to_file(filepath, chunk, offset)
                offset += len(chunk)

    def download_chunks(self, video: "Video", pbar: "tqdm | None" = None) -> int:
        """并发分块下载视频，可选择传入 tqdm 进度条"""
        from concurrent.futures import ThreadPoolExecutor, as_completed

        chunk_size = 10 * 1024 * 1024  # 10MB
        futures: dict = {}
        start = 0
        index = 0

        with ThreadPoolExecutor(max_workers=self.task_num) as executor:
            while start < video.content_len:
                end = min(start + chunk_size, video.content_len) - 1
                if end < start:
                    end = start
                f = executor.submit(self.download_chunk, video, (start, end), index)
                block_size = end - start + 1
                futures[f] = block_size
                start = end + 1
                index += 1

            for f in as_completed(futures):
                f.result()  # 传播异常
                if pbar:
                    pbar.update(futures[f])

        return index

    def download_audio(self, video: "Video", pbar: "tqdm | None" = None) -> None:
        """下载音频流，可选择传入 tqdm 进度条"""
        if not video.audio_url:
            return
        r = self.session.get(video.audio_url, stream=True, timeout=30)
        r.raise_for_status()
        filepath = f"{self.dir}/audio.mp3"
        with open(filepath, "wb") as f:
            for chunk in r.iter_content(chunk_size=8192):
                if chunk:
                    f.write(chunk)
                    if pbar:
                        pbar.update(len(chunk))

    def build_final_video(self, video: "Video", chunk_count: int) -> None:
        """合并分块并调用 ffmpeg 混流"""
        import os
        from .util import mix_video_audio

        video_path = f"{self.dir}/video.{video.format}"

        # 合并所有 chunk
        with open(video_path, "wb") as out:
            for i in range(chunk_count):
                chunk_path = f"{self.dir}/chunk_{i}"
                with open(chunk_path, "rb") as f:
                    out.write(f.read())
                os.remove(chunk_path)

        audio_path = f"{self.dir}/audio.mp3"
        output_path = f"{self.dir}/{video.title}.{video.format}"

        if os.path.exists(audio_path) and os.path.getsize(audio_path) > 0:
            mix_video_audio(video_path, audio_path, output_path)
            os.remove(video_path)
            os.remove(audio_path)
        else:
            os.rename(video_path, output_path)

    def download_danmaku_segment(self, video: "Video", seg_index: int):
        """下载单个弹幕分段"""
        from .danmaku_pb2 import DanmakuSegment

        r = self.session.get(
            self.API_BULLET,
            params={"oid": video.cid, "segment_index": seg_index, "type": 1},
            timeout=30,
        )
        r.raise_for_status()
        segment = DanmakuSegment().parse(r.content)
        return segment

    def download_danmaku(self, video: "Video", pbar: "tqdm | None" = None) -> None:
        """下载并解析弹幕，写入 JSON Lines 文件，可选择传入 tqdm 进度条"""
        from concurrent.futures import ThreadPoolExecutor, as_completed
        import json

        bags = (video.duration + 359) // 360  # 每 6 分钟一个 segment
        results = {}

        with ThreadPoolExecutor(max_workers=self.task_num) as executor:
            futures = {
                executor.submit(self.download_danmaku_segment, video, i + 1): i
                for i in range(bags)
            }
            for f in as_completed(futures):
                idx = futures[f]
                results[idx] = f.result()
                if pbar:
                    pbar.update(1)

        with open(f"{self.dir}/danmuku.txt", "w", encoding="utf-8") as f:
            for i in range(bags):
                for elem in results[i].elems:
                    d = elem.to_dict()
                    f.write(json.dumps(d, ensure_ascii=False) + "\n")

    def run(self, bv: str) -> None:
        """执行完整的下载流程"""
        import os

        video = self.build_video(bv)

        # 创建输出目录
        dir_name = f"{video.title}_{bv}"
        try:
            os.makedirs(dir_name, exist_ok=True)
            self.dir = dir_name
        except OSError:
            os.makedirs(bv, exist_ok=True)
            self.dir = bv

        if video.content_len == 0:
            print(f"download {video.title} fail, video size is 0")
            return

        print(f'download {video.bv} start, title: "{video.title}"')

        chunk_count = self.download_chunks(video)
        self.download_audio(video)
        self.build_final_video(video, chunk_count)
        self.download_danmaku(video)
        print(f'download {video.bv} finished')
